[AArch64][SME] Spill p-regs as z-regs when streaming hazards are possible

MacDue opened the pull request:
This patch adds a new option `-aarch64-enable-zpr-predicate-spills` (which is disabled by default), this option replaces predicate spills with vector spills in streaming[-compatible] functions.

For example:

```
str	p8, [sp, #7, mul vl]            // 2-byte Folded Spill
// ...
ldr	p8, [sp, #7, mul vl]            // 2-byte Folded Reload
```

Becomes:

```
mov	z0.b, p8/z, #1
str	z0, [sp]                        // 16-byte Folded Spill
// ...
ldr	z0, [sp]                        // 16-byte Folded Reload
ptrue	p4.b
cmpne	p8.b, p4/z, z0.b, #0
```

This is done to avoid streaming memory hazards between FPR/vector and predicate spills, which currently occupy the same stack area even when the `-aarch64-stack-hazard-size` flag is set.

This is implemented with two new pseudos SPILL_PPR_TO_ZPR_SLOT_PSEUDO and FILL_PPR_FROM_ZPR_SLOT_PSEUDO. The expansion of these pseudos handles scavenging the required registers (z0 in the above example) and, in the worst case spilling a register to an emergency stack slot in the expansion. The condition flags are also preserved around the `cmpne` in case they are live at the expansion point.

MacDue committed:
[AArch64][SME] Spill p-regs as z-regs when streaming hazards are possible

This patch adds a new option `-aarch64-enable-zpr-predicate-spills`
(which is disabled by default), this option replaces predicate spills
with vector spills in streaming[-compatible] functions.

For example:

```
str	p8, [sp, #7, mul vl]            // 2-byte Folded Spill
// ...
ldr	p8, [sp, #7, mul vl]            // 2-byte Folded Reload
```

Becomes:

```
mov	z0.b, p8/z, #1
str	z0, [sp]                        // 16-byte Folded Spill
// ...
ldr	z0, [sp]                        // 16-byte Folded Reload
ptrue	p4.b
cmpne	p8.b, p4/z, z0.b, #0
```

This is done to avoid streaming memory hazards between FPR/vector and
predicate spills, which currently occupy the same stack area even when
the `-aarch64-stack-hazard-size` flag is set.

This is implemented with two new pseudos SPILL_PPR_TO_ZPR_SLOT_PSEUDO
and FILL_PPR_FROM_ZPR_SLOT_PSEUDO. The expansion of these pseudos
handles scavenging the required registers (z0 in the above example) and,
in the worst case spilling a register to an emergency stack slot in the
expansion. The condition flags are also preserved around the `cmpne`
in case they are live at the expansion point.

MacDue commented on llvm/test/CodeGen/AArch64/spill-fill-zpr-predicates.mir:
```diff
@@ -0,0 +1,1035 @@
+    ; EXPAND-NEXT: $p6 = frame-destroy PTRUE_B 31, implicit $vg
+    ; EXPAND-NEXT: $p6 = frame-destroy CMPNE_PPzZI_B $p6, $z0, 0, implicit-def $nzcv, implicit-def $nzcv
+    ; EXPAND-NEXT: $z0 = frame-destroy LDR_ZXI $sp, 10 :: (load (s128) from %stack.2)
+    ; EXPAND-NEXT: $p5 = frame-destroy PTRUE_B 31, implicit $vg
```
Note: I have a small follow-up patch ready that reduces the amount of `ptrue` instructions created.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -4155,8 +4168,318 @@ int64_t AArch64FrameLowering::assignSVEStackObjectOffsets(
+
+  Register PredReg = AArch64::NoRegister;
+  std::optional<ScopedScavengeOrSpill> FindPPR3bReg;
+  if (AArch64::PPR_3bRegClass.contains(MI.getOperand(0).getReg()))
+    PredReg = MI.getOperand(0).getReg();
+  else
+    FindPPR3bReg.emplace(MF, MBB, MachineBasicBlock::iterator(MI), PredReg,
+                         AArch64::P0, AArch64::PPR_3bRegClass, UsedRegs,
+                         PPR3bRegs, SpillSlots.PPRSpillFI);
```
Unless there's a benefit to use the same register, I believe this if/else is redundant. You should be able to always use `FindPPR3bReg.emplace(..)` directly, because `MI.getOperand(0).getReg()` should be a candidate at this point; it must be dead before MI, otherwise it can't be defined by MI.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -4155,8 +4168,318 @@ int64_t AArch64FrameLowering::assignSVEStackObjectOffsets(
+
+    EmergencyStackSlots SpillSlots;
+    for (MachineBasicBlock &MBB : MF) {
+      for (int Pass = 0; Pass < 2; Pass++) {
```
Please add a comment describing why this needs two passes; and why after two passes all spills should have been replaced.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -4155,8 +4168,318 @@ int64_t AArch64FrameLowering::assignSVEStackObjectOffsets(
+                              .setMemRefs(MI.memoperands())
+                              .getInstr());
+  if (IsNZCVUsed)
+    MachineInstrs.push_back(BuildMI(MBB, MI, DL, TII->get(AArch64::MRS))
+                                .addReg(NZCVSaveReg, RegState::Define)
+                                .addImm(AArch64SysReg::NZCV)
+                                .addReg(AArch64::NZCV, RegState::Implicit)
+                                .getInstr());
```
Is it worth moving this into `ScopedScavengeOrSpill`, so that all registers (Z/P/NZCV) are just handled by this ScopedScavengeOrSpill.

MacDue replied:
I don't think so, we only need to preserve `NZCV` at this point in the fill expansion, so it would not be increasing code reuse to move the `NZCV` logic into  `ScopedScavengeOrSpill`. The `NZCV` logic is also kinda a special case rather than a normal spill/fill.

MacDue commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.td:
```diff
@@ -979,10 +979,19 @@ class ZPRRegOp <string Suffix, AsmOperandClass C, ElementSizeEnum Size,
+
+// Note: This hardware mode is enabled in AArch64Subtarget::getHwModeSet()
+// (without the use of the table-gen'd predicates).
+def SMEWithStreamingMemoryHazards : HwMode<"", [Predicate<"false">]>;
```
The predicate and feature flags of the hardware mode are not used, the implementation in `AArch64Subtarget` override the default implementation (which only checks CPU features). The predicate is only used for hardware-mode specific DAG patterns (of which we have none (https://reviews.llvm.org/D146012).

MacDue replied:
So:
> What if AArch64Subtarget::getHwModeSet would set 1 << 1 instead?

That'd active mode 2 and something would crash, because that does not exist.

sdesmalen-arm replied:
I belief what you're saying, but it's odd to me that TableGen doesn't generate an enum for this, because this means we need to make the implicit assumption that `SMEWithStreamingMemoryHazards == 1`, even though this is not expressed anywhere.

MacDue replied:
I've updated table-gen to make a `AArch64HwModeBits::SMEWithZPRPredicateSpills` enum automatically, which makes this: `return to_underlying(AArch64HwModeBits::SMEWithZPRPredicateSpills);`, which is much less magic :)

MacDue committed:
Fixups

MacDue commented on llvm/lib/Target/AArch64/AArch64Subtarget.cpp:
```diff
@@ -86,6 +86,11 @@ static cl::alias AArch64StreamingStackHazardSize(
     cl::aliasopt(AArch64StreamingHazardSize));
 
+static cl::opt<bool> EnableZPRPredicateSpills(
+    "aarch64-enable-zpr-predicate-spills", cl::init(false), cl::Hidden,
```
I don't think we should right now (possibly with more analysis). If there are streaming memory hazards and predicate spills, this can help. However, I've seen a case were there's only predicate spills (outside the prologue/epilogue), where enabling this will simply replace the predicate reload with a more expensive reload.

MacDue committed:
Fixups

MacDue committed:
Fixups

MacDue committed:
Fixups

MacDue commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -4155,8 +4168,318 @@ int64_t AArch64FrameLowering::assignSVEStackObjectOffsets(
+          TRI.getAllocatableSet(MF, TRI.getRegClass(RegClassID));
+      if (CSRMask)
+        ScavengeableRegs.clearBitsInMask(CSRMask);
+      // TODO: Allow reusing callee-saved registers that have been saved.
```
Done this, still need a little special case for `p4`, but it's simpler now :+1:

MacDue commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -4155,8 +4168,318 @@ int64_t AArch64FrameLowering::assignSVEStackObjectOffsets(
+  assert((!Spilled || !InPrologueOrEpilogue) &&
+         "FILL_PPR_FROM_ZPR_SLOT_PSEUDO expansion should not spill in prologue "
+         "or epilogue");
+#endif
```
Done (via `SpillSlots` pointer) :+1:

MacDue commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -4155,8 +4168,318 @@ int64_t AArch64FrameLowering::assignSVEStackObjectOffsets(
+    EmergencyStackSlots SpillSlots;
+    for (MachineBasicBlock &MBB : MF) {
+      for (int Pass = 0; Pass < 2; Pass++) {
+        bool HasPPRSpills = expandSMEPPRToZPRSpillPseudos(
```
Split this into `ScavengeableRegsFrameSetup` and `ScavengeableRegsBody` (which are computed early on, then chosen based on the `MI` flags).

MacDue committed:
Remove overly cautious code

Turns out LiveRegUnits will already reserve unsaved callee-saved
registers so we don't need to worry about doing this here.

MacDue committed:
Fixups

MacDue commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -5200,9 +5501,14 @@ void AArch64FrameLowering::emitRemarks(
+            if (MI.getOpcode() != AArch64::SPILL_PPR_TO_ZPR_SLOT_PSEUDO &&
+                MI.getOpcode() != AArch64::FILL_PPR_FROM_ZPR_SLOT_PSEUDO &&
+                AArch64::PPRRegClass.contains(MI.getOperand(0).getReg())) {
+              MI.dump();
```
Yep :+1:

MacDue commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -4155,8 +4168,297 @@ int64_t AArch64FrameLowering::assignSVEStackObjectOffsets(
+      static_cast<const AArch64InstrInfo *>(MF.getSubtarget().getInstrInfo());
+
+  Register ZPredReg = AArch64::NoRegister;
+  ScopedScavengeOrSpill FindZPRReg(
```
Sure, I like that idea :+1:

MacDue committed:
Tweak

sdesmalen-arm reviewed (approved):
LGTM, thanks @MacDue
